File: LeetCode/Problems/0088_Merge_Sorted_Array/merge_sorted_array.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        return solution(nums1, m, nums2, n);
    }

private:
    // time complexity: O(n+m)
    // space complexity: O(1)
    void solution(vector<int>& nums1, int m, vector<int>& nums2, int n){
        int i = m - 1;
        int j = n - 1;
        int k = m + n - 1;

        while(j >= 0)
            nums1[k--] = (i >= 0 && nums1[i] > nums2[j]) ? nums1[i--] : nums2[j--];
    }
};

// nums is passed by constant reference and cannot be changed by this function
string printVector(const vector<int>& nums){
    string str;

    str.append("[");
    for(const auto& n : nums){
        if(n != *nums.begin())
            str.append(", ");
        str.append(std::to_string(n));
    }
    str.append("]");

    return str;
}

int main(){
    int m = 3, n = 3;
    vector<int> nums1 = {1,2,3,0,0,0};
    vector<int> nums2 = {2,5,6};

    cout << "Input: \n";
    cout << printVector(nums1) << ",  m = " << m << "\n";
    cout << printVector(nums2) << ",  n = " << n << "\n";

    Solution sol;
    sol.merge(nums1, m, nums2, n);
    cout << "\nOutput: \n" << printVector(nums1) << "\n";

    return 0;
}

File: LeetCode/Problems/0088_Merge_Sorted_Array/merge_sorted_array.py
from typing import List

class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        """
        Do not return anything, modify nums1 in-place instead.
        """

        i = m - 1
        j = n - 1
        k = m + n - 1

        while j >= 0:
            if i >= 0 and nums1[i] > nums2[j]:
                nums1[k] = nums1[i]
                i -= 1
            else:
                nums1[k] = nums2[j]
                j -= 1
            k -= 1

def printList(nums: List[int]) -> str:
    return ("[" + (', ').join(map(str, nums)) + "]")

nums1 = [1,2,3,0,0,0]
nums2 = [2,5,6]
m = 3
n = 3
print("Input: ")
print(printList(nums1) + ", m = " + str(m))
print(printList(nums2) + ", n = " + str(n))

Solution().merge(nums1, 3, nums2, 3)
print("\nOutput: ")
print(nums1)
